Foreground notification service

Eyenseo opened the pull request:
This adds an option to run a service in the foreground to keep the app from being destroyed. After updating to Oreo the app missed a lot of notifications - before Oreo I didn't had any problems as the app was pretty responsive / wasn't destroyed.  The service fixes this problem as it keeps the app running.
Additionally I increased the target sdk to 28 as this allows the user to hide the notification dot from the system bar (manually).

I'm not an android developer so there might be some things that can be solved easier - feedback is welcome.
I tested this patch set on CarbonROM / Android 8.1.0 and can't vouch how other versions will behave.

rootkiwi commented:
Nice! I just switched to LineageOS 15.1 (oreo 8.1) also, will look into this tomorrow!

rootkiwi commented:
I think it looks great.

Though Android docs is a little unclear about this new Background Execution Limits in Oreo.

src: https://developer.android.com/about/versions/oreo/background#services
> Another foreground app is connected to the app, either by binding to one of its services or by making use of one of its content providers. For example, the app is in the foreground if another app binds to its: 
> - Notification listener
> - ...

So what does that mean? AN2Linux have a `NotificationListenerService` that the system binds to, is the system considered as a foreground app? What is the implications for a `NotificationListenerService` with Android Oreo and the new Background Limits, does the system unbind the `NotificationListenerService` now because of Background Limits after a while? I don't know...

But as you said, you have noticed that AN2Linux started missing notifications with Oreo.

So this feature is then needed, and it's even an opt-in feature so the user of AN2Linux can choose to start a foreground service if it's needed.

rootkiwi commented:
Also I'm not sure if the system should start the forground service automatically on system boot, if the user have enabled the useForegroundService.

rootkiwi commented on app/src/main/java/kiwi/root/an2linuxclient/activities/MainSettingsActivity.java:
```diff
@@ -19,32 +18,49 @@
+            startService(new Intent(this, AN2LinuxService.class));
+        }
+    }
 
```
I think we could use the SharedPreferences that's automatically created here for if the user have enabled useForegroundService or not.

Like this instead:

```
boolean useForegroundService = PreferenceManager.getDefaultSharedPreferences(this)
        .getBoolean(getString(R.string.preference_enable_service), false);
if (useForegroundService){
    startService(new Intent(this, AN2LinuxService.class));
}
```

And also I think the code in `onStart`:
```
if (useForegroundService){
    startService(new Intent(this, AN2LinuxService.class));
}
```
is not needed, `onResume` is enough as it will be called every time after `onCreate` and `onStart` including the first time.

Eyenseo commented:
I addressed the feedback and added a boot receiver that starts the foreground service if the foreground service is enabled as well as the application.  I also improved the ux a little bit.

rootkiwi commented:
Great additions!

Now just one more thing. The `Enable foreground service` setting is now disabled if the user exits and then opens an2linux, with disabled I mean the user is not able to change it. Is there a reason for that? It means that if the user (after exiting and opening again) wants to disable the foreground service setting the user needs to disable and enable the `Enable AN2Linux` setting to let the enable code run:
https://github.com/rootkiwi/an2linuxclient/pull/14/commits/bfba5d07219cd2be7683d4ef6dc32aa28f603b7a#diff-44de7e5cf7e74d74bb24cead69bbb41eR164
```
findPreference(getString(R.string.preference_enable_service)).setEnabled(true);
```

If I understand correctly, the goal is to have the `Enable foreground service` setting enabled (allow user input) only if the user have enabled `Enable AN2Linux`? Then we could remove these two lines:
```
findPreference(getString(R.string.preference_enable_service)).setEnabled(true);
```
and
```
findPreference(getString(R.string.preference_enable_service)).setEnabled(false);
```

And then edit this line:
https://github.com/rootkiwi/an2linuxclient/commit/bfba5d07219cd2be7683d4ef6dc32aa28f603b7a#diff-058506910acabadc814420237b486a55R14
```
android:enabled="false"
```
to
```
android:dependency="@string/preference_enable_an2linux"
```

which automatically will have the `Enable foreground service` only allowing user input if `Enable AN2Linux` is enabled.

rootkiwi commented:
Great, thank you very much :+1:

Merging (rebasing) now and will update the app in a while as well.
